driver-core/src/main/com/mongodb/connection/SplittablePayloadCommandMessage.java

```diff
@@ -0,0 +1,165 @@
+import org.bson.BsonDocument;
+import org.bson.BsonString;
+import org.bson.BsonValue;
+import org.bson.BsonWriterSettings;
+import org.bson.FieldNameValidator;
+import org.bson.codecs.BsonValueCodecProvider;
+import org.bson.codecs.Codec;
+import org.bson.codecs.EncoderContext;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.io.BsonOutput;
+
+import java.util.Map;
+
+import static org.bson.codecs.configuration.CodecRegistries.fromProviders;
+
+/**
+ * A command message that uses OP_MSG or OP_QUERY to send the command.
+ */
+class SplittablePayloadCommandMessage extends CommandMessage {
+    private static final CodecRegistry REGISTRY = fromProviders(new BsonValueCodecProvider());
+    private static final FieldNameValidator NOOP_FIELD_NAME_VALIDATOR = new NoOpFieldNameValidator();
+    private static final int HEADROOM = 16 * 1024;
+    private final MongoNamespace namespace;
+    private final BsonDocument command;
+    private final SplittablePayload payload;
+
+    SplittablePayloadCommandMessage(final MongoNamespace namespace, final BsonDocument command, final SplittablePayload payload,
+                                    final MessageSettings settings) {
+        super(namespace.getFullName(), getOpCode(settings), settings);
+        this.namespace = namespace;
+        this.command = command;
+        this.payload = payload;
+    }
+
+    @Override
+    protected EncodingMetadata encodeMessageBodyWithMetadata(final BsonOutput bsonOutput, final SessionContext sessionContext) {
+        if (useOpMsg()) {
+            bsonOutput.writeInt32(0);  // flag bits
+            bsonOutput.writeByte(0);   // payload type
+
+            int commandStartPosition = bsonOutput.getPosition();
+
+            BsonBinaryWriter commandWriter = new BsonBinaryWriter(new BsonWriterSettings(),
+                    new BsonBinaryWriterSettings(getSettings().getMaxDocumentSize()), bsonOutput, NOOP_FIELD_NAME_VALIDATOR);
+
+            commandWriter.writeStartDocument();
+            encodeCommand(command, commandWriter);
+            encodeBsonValue("$db", new BsonString(namespace.getDatabaseName()), commandWriter);
+            if (sessionContext.hasSession()) {
+                if (sessionContext.getClusterTime() != null) {
+                    encodeBsonValue("$clusterTime", sessionContext.getClusterTime(), commandWriter);
+                }
+                encodeBsonValue("lsid", sessionContext.getSessionId(), commandWriter);
+            }
+            commandWriter.writeEndDocument();
+
+            bsonOutput.writeByte(1);   // payload type
+            int sectionPosition = bsonOutput.getPosition();
+            bsonOutput.writeInt32(0); // size
+            bsonOutput.writeCString(payload.getPayloadName());
+            BsonBinaryWriter payloadWriter = new BsonBinaryWriter(new BsonWriterSettings(),
+                    new BsonBinaryWriterSettings(getSettings().getMaxDocumentSize() + HEADROOM), bsonOutput,
```

Comment:
HEADROOM maybe too big here, but we have tests expecting slightly larger documents (Request operations) to be successful and I couldn't determine what would be an acceptable amount.